File: java/amp-eventing/src/main/java/amp/eventing/streaming/StreamingHeadersProcessor.java
package amp.eventing.streaming;

import amp.messaging.EnvelopeHelper;
import amp.messaging.IContinuationCallback;
import amp.messaging.IMessageProcessor;
import amp.messaging.MessageContext;
import amp.messaging.MessageException;

public class StreamingHeadersProcessor implements IMessageProcessor {
    @Override
    public void processMessage(MessageContext context, IContinuationCallback continuation) throws MessageException  {
        final String endOfStreamTopic = EndOfStream.class.getCanonicalName();
        final String collectionSizeTopic = CollectionSizeNotifier.class.getCanonicalName();

        EnvelopeHelper env = new EnvelopeHelper(context.getEnvelope());

        if (context.getDirection().equals(MessageContext.Directions.Out)) {

            if (env.getMessageTopic().equals(endOfStreamTopic)) {
                env.setMessageTopic(((EndOfStream)context.getMessage()).getStreamType());
            }

            if (env.getMessageTopic().equals(collectionSizeTopic)) {
                env.setMessageTopic(((CollectionSizeNotifier)context.getMessage()).getCollectionType());
            }
        }

        continuation.continueProcessing();
    }
}

File: dotnet/source/amp.eventing/streaming/StreamingHeadersProcessor.cs
using amp.messaging;
using cmf.bus;
using System;

namespace amp.eventing.streaming
{
    public class StreamingHeadersProcessor : IMessageProcessor
    {
        public void ProcessMessage(MessageContext context, Action continueProcessing)
        {
            string endOfStreamTopic = typeof(EndOfStream).FullName;
            string collectionSizeTopic = typeof(CollectionSizeNotifier).FullName;

            if (context.Direction == MessageContext.Directions.Out)
            {
                if (context.Envelope.GetMessageTopic() == endOfStreamTopic)
                {
                    context.Envelope.SetMessageTopic(((EndOfStream)context.Message).StreamType);
                }

                if (context.Envelope.GetMessageTopic() == collectionSizeTopic)
                {
                    context.Envelope.SetMessageTopic(((CollectionSizeNotifier)context.Message).CollectionType);
                }
            }

            continueProcessing();
        }
    
        public void Dispose()
        {
            //Nothing to do.
        }
    }
}
